File: lib/items/texture_pages.rb
require_relative 'base'

TEXTURE_PAGE_FIELDS = {
  base: [
    { key: :width, type: :int },
    { key: :height, type: :int },
    { key: :id, type: :int }
  ],
  box: [
    { key: :x0, type: :int },
    { key: :y0, type: :int },
    { key: :x2, type: :int },
    { key: :y2, type: :int }
  ]
}.freeze

module Wld
  module Items
    class TexturePages < Base
      def initialize(file_or_hash, marker = 'TEXP')
        @separator = 'PAGE'
        super
      end

      protected

      def unpack_node(file, _index)
        file.skip # always 2, skip
        page = file.read TEXTURE_PAGE_FIELDS[:base]
        textures_count = file.int
        page[:textures] = textures_count.times.map { unpack_texture(file) }

        token = file.word
        raise StandardError, "Not found TXPG separator (got '#{token}')" unless token == 'TXPG'

        page[:is_alpha] = file.bool
        page[:image_binary] = file.hex(page[:width] * page[:height] * 2) # the dds image binary data

        page
      end

      def pack_node(info)
        accumulator = BinaryDataBuffer.new

        accumulator.push_int 2
        accumulator.add TEXTURE_PAGE_FIELDS[:base], info
        accumulator.push_int info[:textures].size
        accumulator.concat pack_texture_list(info[:textures])
        accumulator.push_word('TXPG')
        accumulator.push_bool(info[:is_alpha])
        accumulator.concat info[:image_binary]

        accumulator.data
      end

      def unpack_texture(file)
        {
          filepath: file.filename,
          box: file.read(TEXTURE_PAGE_FIELDS[:box]),
          source_box: file.read(TEXTURE_PAGE_FIELDS[:box])
        }
      end

      def pack_texture_list(textures)
        accumulator = BinaryDataBuffer.new
        textures.each do |texture|
          accumulator.push_name texture[:filepath]
          accumulator.add TEXTURE_PAGE_FIELDS[:box], texture[:box]
          accumulator.add TEXTURE_PAGE_FIELDS[:box], texture[:source_box]
        end
        accumulator.data
      end
    end
  end
end

File: unity_scripts/Assets/Editor/WldImporter/WldParser/Items/TexturePages.cs
using System;
using WldParser.Helpers;
using System.Collections.Generic;

namespace WldParser.Items
{
    [Serializable]
    public class TextureBox
    {
        public int x0, y0, x2, y2;
        public int Width => x2 - x0;
        public int Height => y2 - y0;

        public TextureBox(int x0, int y0, int x2, int y2)
        {
            this.x0 = x0; this.y0 = y0; this.x2 = x2; this.y2 = y2;
        }
    }

    [Serializable]
    public class SubTextureInfo
    {
        public string filepath;
        public TextureBox box;
        public TextureBox sourceBox;
    }

    [Serializable]
    public class TexturePageInfo
    {
        public int width;
        public int height;
        public int id;
        public List<SubTextureInfo> textures = new List<SubTextureInfo>();
        public bool isAlpha;
        public byte[] imageBinary;
    }
}

namespace WldParser.Items
{
    public class TexturePages : WldSectionBase<TexturePageInfo>
    {
        // Передаем в базовый конструктор маркер "TEXP" и разделитель "PAGE"
        public TexturePages(BinaryFileReader file) : base(file, "TEXP", "PAGE")
        {
        }

        protected override TexturePageInfo UnpackNode(BinaryFileReader file, int index, int entrySize)
        {
            var page = new TexturePageInfo();

            // 1. Ruby: file.skip # always 2
            // В Ruby скрипте pack_node делает push_int 2, значит читаем 4 байта (int).
            int unknownVersion = file.Int(); 

            // 2. Ruby: page = file.read TEXTURE_PAGE_FIELDS[:base]
            page.width = file.Int();
            page.height = file.Int();
            page.id = file.Int();

            // 3. Ruby: textures_count = file.int
            int texturesCount = file.Int();

            // 4. Ruby: page[:textures] = ... map { unpack_texture(file) }
            for (int i = 0; i < texturesCount; i++)
            {
                page.textures.Add(UnpackTexture(file));
            }

            // 5. Ruby: token = file.word; raise ... unless token == 'TXPG'
            // ВАЖНО: TXPG находится внутри блока PAGE, а не является разделителем блоков.
            string token = file.Word();
            if (token != "TXPG")
            {
                // Используем ваш класс исключения, хоть он и для внешней структуры,
                // но здесь он тоже подходит по смыслу.
                throw new TokenError(token, "TXPG");
            }

            // 6. Ruby: page[:is_alpha] = file.bool
            page.isAlpha = file.Bool();

            // 7. Ruby: page[:image_binary] = file.hex(page[:width] * page[:height] * 2)
            // Формат пикселей 16-битный (2 байта на пиксель)
            int dataSize = page.width * page.height * 2;
            page.imageBinary = file.Raw(dataSize);

            return page;
        }

        private SubTextureInfo UnpackTexture(BinaryFileReader file)
        {
            var info = new SubTextureInfo();

            // Ruby: file.filename
            info.filepath = file.Name();

            // Ruby: file.read(TEXTURE_PAGE_FIELDS[:box])
            info.box = new TextureBox(file.Int(), file.Int(), file.Int(), file.Int());

            // Ruby: file.read(TEXTURE_PAGE_FIELDS[:box]) (source_box)
            info.sourceBox = new TextureBox(file.Int(), file.Int(), file.Int(), file.Int());

            return info;
        }
    }
}